doc/designs/ech-api.md

```diff
@@ -0,0 +1,718 @@
+
+To access the `OSSL_ECHSTORE` associated with an `SSL_CTX` or
+`SSL` connection:
+
+```c
+OSSL_ECHSTORE *SSL_CTX_get1_echstore(const SSL_CTX *ctx);
+OSSL_ECHSTORE *SSL_get1_echstore(const SSL *s);
+```
+The resulting `OSSL_ECHSTORE` can be modified and then re-associated
+with an `SSL_CTX` or `SSL` connection.
+
+Finer-grained client control
+----------------------------
+
+TODO(ECH): revisit this later, when we hopefully have some more information
+about ECH deployments.
+
+Applications that need fine control over which ECHConfigList (from those
+available) will be used, can query an `OSSL_ECHSTORE`, retrieving information
+about the set of "singleton" ECHConfigList values available, and then, if
+desired, down-select to one of those, e.g., based on the `public_name` that
+will be used. This would enable a client that selects the server address to use
+based on IP address hints that can also be present in an HTTPS/SCVB resource
+record to ensure that the correct matching ECH public value is used. The
+information is presented to the caller using the `OSSL_ECH_INFO` type, which
+provides a simplified view of ECH data, but where each element of an array
+corresponds to exactly one ECH public value and set of names.
+
+```c
+/*
+ * Application-visible form of ECH information from the DNS, from config
+ * files, or from earlier API calls. APIs produce/process an array of these.
+ */
+typedef struct ossl_ech_info_st {
+    int index; /* externally re-usable reference to this value */
+    char *public_name; /* public_name from API or ECHConfig */
+    char *inner_name; /* server-name (for inner CH if doing ECH) */
+    unsigned char *outer_alpns; /* outer ALPN string */
+    int outer_alpns_len;
+    unsigned char *inner_alpns; /* inner ALPN string */
+    int inner_alpns_len;
+    char *echconfig; /* a JSON-like version of the associated ECHConfig */
+} OSSL_ECH_INFO;
+
+void OSSL_ECH_INFO_free(OSSL_ECH_INFO *info, int count);
+int OSSL_ECH_INFO_print(BIO *out, OSSL_ECH_INFO *info, int count);
+```
+
+ECH Store Internals
+-------------------
+
+The internal structure of an ECH Store is as described below:
+
+```c
+typedef struct ossl_echext_st {
+    unsigned int type;
+    unsigned int len;
+    unsigned char *val;
+} OSSL_ECHEXT;
+
+DEFINE_STACK_OF(OSSL_ECHEXT)
+
+typedef struct ossl_echstore_entry_st {
+    unsigned int version; /* 0xff0d for draft-13 */
+    char *public_name;
+    unsigned int pub_len;
+    unsigned char *pub;
+    unsigned int nsuites;
+    OSSL_HPKE_SUITE *suites;
+    unsigned int max_name_length;
+    uint8_t config_id;
+    STACK_OF(OSSL_ECHEXT) *exts;
+    char *pemfname; /* name of PEM file from which this was loaded */
+    time_t loadtime; /* time public and private key were loaded from file */
+    EVP_PKEY *keyshare; /* long(ish) term ECH private keyshare on a server */
+    int for_retry; /* whether to use this ECHConfigList in a retry */
+    unsigned int encoded_len; /* length of overall encoded content */
```

Comment:
Here and in a lot of place, you really need to define a 'byte buffer' / 'byte string' to carry and manage more easily all these buffers.
Just a simple and fine wrapper to simplify buffers handling.

Something like : 
`````
typedef struct short_buffer_st {
    uchar_t   *data;
    uint32_t   len;
} OSSL_SHRT_BUF;

typedef struct byte_buffer_st {
    uchar_t  *data;
    size_t      len;
} OSSL_BYTEBUF;

/* plus some TYPE_new() TYPE_free() methods */

`````

Reply:
I thought about that, but such a generic change really ought be it's own PR was my conclusion as it's independent of ECH. Should the project adopt such structures for internal or internal and external use, I'd be fine to adopt those. (I guess the argument against is that to some extent, the very large set of changes to nearly-fully adopt such structures would seem like a lot of work for very little return.)

So, I'd prefer leave this as-is for now.